sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -279,7 +283,9 @@ case class StaticInvoke(
     inputTypes: Seq[AbstractDataType] = Nil,
     propagateNull: Boolean = true,
     returnNullable: Boolean = true,
-    isDeterministic: Boolean = true) extends InvokeLike {
+    isDeterministic: Boolean = true,
+    scalarFunctionName: Option[String] = None,
```

Comment:
Hmm instead of adding these two parameters, can we instead check `staticObject` and `functionName`?

Reply:
+1, we can check if `staticObject` is a subclass of `BoundFunction`

Reply:
For V2 UDF, the `staticObject` here is the class of the `BoundFunction`, and the `functionName` is the magic name: `invoke`. However, we need the `name` and `canonicalName` to build the `UserDefinedScalarFunc`.

Reply:
we can instantiate `staticObject` and call its `canonicalName` function?

Reply:
Seems is not easy. The `BoundFunction` is created by calling the bind method of `UnboundFunction` which is loaded by `FunctionCatalog`. And we have no guarantee the `BoundFunction` has no-args constructor to create with reflect.

Reply:
Maybe we can take a try with `TreeNodeTag` to save those two names. This can avoid breaking the `StaticInvoke` constructor. However, I am not sure it is a good idea.

Reply:
Yea looks like it's not easy to re-use `StaticInvoke`. Instead of stacking the extra parameters on top of the class though, I wonder if it makes sense to create another class such as `V2StaticInvoke` or something for this specific use case. Just a nit.

Reply:
I am fine with a dedicated class. There may be some redundant code.

Reply:
It's better to reuse `StaticInoke` so that any optimizations for it can still apply. How about we add one more parameter `function: Option[ScalarFunction[_]]` to it instead of two strings?

Reply:
I was thinking extending `StaticInvoke` somehow like
```scala
case class V2StaticInvoke(wrapped: StaticInvoke, name: String, canonicalName: String) extends InvokeLike
```
but this way we still need to override quite a few members in `InvokeLike` so may not worth it.

Reply:
> How about we add one more parameter function: Option[ScalarFunction[_]] to it instead of two strings?

Changed to this way. And it is the initial implementation.

Reply:
@cloud-fan @sunchao Although adding `ScalarFunction` as a new parameter is convenient, it makes StaticInvoke look very strange.
We already call the `scalarFunc.getClass`, `scalarFunc.resultType()`, `scalarFunc.isResultNullable` and `scalarFunc.isDeterministic` before passed the new `ScalarFunction`.

Reply:
I'm OK to keep as it is. For me it is just a nit and one parameter makes the change less intrusive.